File: golang/leetcode/python/generate_parentheses.py
class LeetCode_GenerateParentheses(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n == 0:
            return []
        res = []
        self.dfs(n, n, '', res)
        return res

    def dfs(self, left, right, path, res):
        if left == 0 and right == 0:
            res.append(path)
        if left > 0:
            self.dfs(left - 1, right, path + '(', res)
        if right > left:
            self.dfs(left, right - 1, path + ')', res)

def main():
    leetcode = LeetCode_GenerateParentheses()
    print(leetcode.generateParenthesis(3))
    

File: golang/leetcode/generate_parentheses.go
package leetcode

// https://leetcode.com/problems/generate-parentheses/
//
// Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
//
// For example, given n = 3, a solution set is:
//
// [
//   "((()))",
//   "(()())",
//   "(())()",
//   "()(())",
//   "()()()"
// ]

func generateParenthesis(n int) []string {
	res := []string{}
	if n == 0 {
		return res
	}
	backtrack(&res, "", 0, 0, n)
	return res
}

func backtrack(res *[]string, s string, open, close, max int) {
	if len(s) == max*2 {
		*res = append(*res, s)
		return
	}
	if open < max {
		backtrack(res, s+"(", open+1, close, max)
	}
	if close < open {
		backtrack(res, s+")", open, close+1, max)
	}
}